src/plonk_challenger.rs

```diff
@@ -1,24 +1,39 @@
 use std::marker::PhantomData;
 
-use crate::{rescue_sponge, AffinePoint, AffinePointTarget, CircuitBuilder, Curve, Field, HaloCurve, ProjectivePoint, Target};
+use crate::{AffinePoint, AffinePointTarget, CircuitBuilder, Curve, Field, HaloCurve, ProjectivePoint, rescue_permutation, RESCUE_SPONGE_RATE, RESCUE_SPONGE_WIDTH, Target};
 
 /// Observes prover messages, and generates challenges by hashing the transcript.
 #[derive(Clone)]
 pub(crate) struct Challenger<F: Field> {
-    transcript: Vec<F>,
+    sponge_state: Vec<F>,
+    input_buffer: Vec<F>,
+    output_buffer: Vec<F>,
     security_bits: usize,
 }
 
+/// Observes prover messages, and generates verifier challenges based on the transcript.
+///
+/// The implementation is roughly based on a duplex sponge with a Rescue permutation. Note that in
+/// each round, our sponge can absorb an arbitrary number of prover messages and generate an
+/// arbitrary number of verifier challenges. This might appear to diverge from the duplex sponge
+/// design, but it can be viewed as a duplex sponge whose inputs are sometimes zero (when we perform
+/// multiple squeezes) and whose outputs are sometimes ignored (when we perform multiple
+/// absorptions). Thus the security properties of a duplex sponge still apply to our design.
```

Comment:
This works fine provided that the number of absorptions and squeezes in each round is fixed. Otherwise you need explicit padding, which I don't think your implementation has.

Reply:
Hm I'm not seeing why a non-uniform round structure would be an issue; could you elaborate on that? This should be equivalent to a duplex modulo input padding and extra (ignored) outputs. E.g. if a particular round has `r` inputs and `2r` outputs, we can view it as adding two duplexes, where the second duplex gets an input of zero.

Reply:
My point is just that the input to the hash has to be unambiguously parseable as a transcript of the interactive protocol, which might not be the case if you don't use padding and have a variable number of absorptions or squeezes in any round.

Reply:
I think I see what you mean. If the prover can influence the length of the transcript, then our (implicit) padding function is not injective, since `pad(m) = pad(m || 0)` if `m`'s length is not a multiple of `r`. Is that the issue?

Currently the number of prover messages in a given round is fixed (for a fixed instance) for us, but I agree that we would need a different padding function like `pad10*` if we had a variable-length round.

Reply:
Thanks for taking a look!